sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ArrayBasedMapData.scala

```diff
@@ -35,6 +37,29 @@ class ArrayBasedMapData(val keyArray: ArrayData, val valueArray: ArrayData) exte
   override def toString: String = {
     s"keys: $keyArray, values: $valueArray"
   }
+
+  override def equals(obj: Any): Boolean = {
+    if (obj == null && this == null) {
+      return true
+    }
+
+    if (obj == null || !obj.isInstanceOf[ArrayBasedMapData]) {
+      return false
+    }
+
+    val other = obj.asInstanceOf[ArrayBasedMapData]
+
+    keyArray.equals(other.keyArray) && valueArray.equals(other.valueArray)
+  }
```

Comment:
I believe this part is also causing `ComplexDataSuite` to fail the "inequality test for MapData" test case, where we use the triple inequality check for object equality. That should be calling `equalsTo`?  This PR's implementation of `equals` follows the expectation of `Literal`. Can you comment on the expected behaviour for `equals()`, @cloud-fan ? Is it supposed to be object equality or element equality?

Reply:
ideally `equals` should do element equality.

Reply:
Not sure how we do it in other code, but there is Arrays.deepEquals for element-wise equality